pkg/storegateway/batch_series_test.go

```diff
@@ -771,52 +773,169 @@ func TestBlockSeriesChunkRefsSetsIterator_ErrorPropagation(t *testing.T) {
+					{lset: labels.FromStrings("a", "2", "b", "2"), chunks: make([]seriesChunkRef, 1)},
+				}},
+			},
+		},
+		"selects all series in multiple batches": {
+			matcher:     labels.MustNewMatcher(labels.MatchRegexp, "a", ".+"),
+			batchSize:   1,
+			chunksLimit: 100,
+			seriesLimit: 100,
+			expectedSeries: []seriesChunkRefsSet{
+				{series: []seriesChunkRefs{
+					{lset: labels.FromStrings("a", "1", "b", "1"), chunks: make([]seriesChunkRef, 1)},
+				}},
+				{series: []seriesChunkRefs{
+					{lset: labels.FromStrings("a", "1", "b", "2"), chunks: make([]seriesChunkRef, 1)},
+				}},
+				{series: []seriesChunkRefs{
+					{lset: labels.FromStrings("a", "2", "b", "1"), chunks: make([]seriesChunkRef, 1)},
+				}},
+				{series: []seriesChunkRefs{
+					{lset: labels.FromStrings("a", "2", "b", "2"), chunks: make([]seriesChunkRef, 1)},
+				}},
+			},
+		},
+		"selects some series in single batch": {
+			matcher:     labels.MustNewMatcher(labels.MatchEqual, "a", "1"),
+			batchSize:   100,
+			chunksLimit: 100,
+			seriesLimit: 100,
+			expectedSeries: []seriesChunkRefsSet{
+				{series: []seriesChunkRefs{
+					{lset: labels.FromStrings("a", "1", "b", "1"), chunks: make([]seriesChunkRef, 1)},
+					{lset: labels.FromStrings("a", "1", "b", "2"), chunks: make([]seriesChunkRef, 1)},
+				}},
+			},
+		},
+		"selects some series in multiple batches": {
+			matcher:     labels.MustNewMatcher(labels.MatchEqual, "a", "1"),
+			batchSize:   1,
+			chunksLimit: 100,
+			seriesLimit: 100,
+			expectedSeries: []seriesChunkRefsSet{
+				{series: []seriesChunkRefs{
+					{lset: labels.FromStrings("a", "1", "b", "1"), chunks: make([]seriesChunkRef, 1)},
+				}},
+				{series: []seriesChunkRefs{
+					{lset: labels.FromStrings("a", "1", "b", "2"), chunks: make([]seriesChunkRef, 1)},
+				}},
+			},
+		},
 	}
-	suite.cache.SwapWith(noopCache{})
-
-	indexReader := firstBlock.indexReader()
-	defer indexReader.Close()
-
-	matcher, err := labels.NewMatcher(labels.MatchRegexp, "a", ".+")
-	require.NoError(t, err)
-
-	iterator, err := openBlockSeriesChunkRefsSetsIterator(
-		ctx,
-		100,
-		indexReader,
-		firstBlock.meta.ULID,
-		[]*labels.Matcher{matcher},
-		nil,
-		suite.store.seriesHashCache.GetBlockCache(firstBlock.meta.ULID.String()),
-		&limiter{limit: 1},
-		&limiter{limit: 100000},
-		false,
-		firstBlock.meta.MinTime,
-		firstBlock.meta.MaxTime,
-		nil,
-		newSafeQueryStats(),
-		suite.logger,
-	)
-	require.NoError(t, err)
 
-	_ = readAllSeriesChunkRefsSet(iterator)
-	assert.ErrorContains(t, iterator.Err(), "test limit exceeded")
+	for testName, testCase := range testCases {
+		testName, testCase := testName, testCase
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+
+			suite := prepareStoreWithTestBlocks(t, t.TempDir(), objstore.NewInMemBucket(), false, NewChunksLimiterFactory(0), NewSeriesLimiterFactory(0))
+			var firstBlock *bucketBlock
+			// Find the block with the smallest timestamp in its ULID
+			for _, b := range suite.store.blocks {
```

Comment:
A couple of comments:
1. Why do we need the 1st block by time? It's unclear to me. I would improve the comment to explain the why.
2. You may move this logic as a function of `suite`

Reply:
> Why do we need the 1st block by time? It's unclear to me. I would improve the comment to explain the why.

because the series are split by block - first 4 series go to the block that's created first, second 4 go to the block that's created after. And block's time is also `time.Now()` each time the test runs, so the ULID is dynamic. Only order between ULID's time is static. It's quite the kludge, I agree.

> You may move this logic as a function of suite

I want to get rid of this test setup after the refactor on `blockSeriesChunkRefsSetsIterator`. I'd rather keep the code here, so it's easier to clean up


--- 

I left a comment with a TODO and an explanation

Reply:
👍 Ok, thanks for the details!